docs/source/content/api.md

```diff
@@ -73,6 +71,50 @@ the format of valid observations is specified by `env.observation_space`.
 In the example above we sampled random actions via `env.action_space.sample()`. Note that we need to seed the action space separately from the 
 environment to ensure reproducible samples.
 
+### Change in env.step API
+
+Previously, the step method returned only one boolean - `done`. This is being deprecated in favour of returning two booleans `terminated` and `truncated`.
+
+Environment termination can happen due to any number of reasons inherent to the environment eg. task completition, failure etc. This is distinctly different from an episode of the environment ending due to a user-set time-limit, referred to as 'truncation'. 
+
+`terminated` signal is set to `True` when the core environment terminates inherently because of task completition, failure etc. 
+`truncated` signal is set to `True` when the episode ends specifically because of a time-limit not inherent to the environment. 
+
+It is possible for `terminated=True` and `truncated=True` to occur at the same time when termination and truncation occur at the same step. 
+
+#### Motivation 
+This is done to remove the ambiguity in the `done` signal. `done=True` does not distinguish between the environment terminating and the episode truncating. This problem was avoided previously by setting `info['TimeLimit.truncated']` in case of a timelimit through the TimeLimit wrapper. However, this forces the truncation to happen only through a wrapper, and does not accomodate environments which truncate in the core environment. 
+
+The main reason we need to distinguish between these two signals is the implications in implementing the Temporal Difference update, where the reward along with the value of the next state(s) are used to update the value of the current state. At terminal states, there is no contribution from the value of the next state(s) and we only consider the value as a function of the reward at that state.  However, this is not true when we forcibly truncate the episode. Time-limits are often set during training to diversify experience, but the goal of the agent is not to maximize reward over this time period. The `done` signal is often used to control whether the value of the next state 
```

Comment:
The main difference is more high-level than TD update: it is whether we treat the task as a finite or infinite horizon one.
See https://github.com/DLR-RM/stable-baselines3/blob/master/tests/test_gae.py#L136 for concrete example and test.

Reply:
> where basically terminated is always False 

Terminated could be true (for instance in case of a failure), but I would prefer to talk about termination reasons (see my comments in https://github.com/openai/gym/issues/2510#issuecomment-1099989405), as in practice, an episode is terminated if you reach the timeout.

> bootstrapping next state value is skipped leading to incorrect value estimates,

yes, unless you pass the remaining time as input (as described in https://arxiv.org/abs/1712.00378, what you call "time-aware" below).

> suppose if the task timelimit is 1000 steps, and the agent is aware of this, and the user sets an additional episode time limit 

I'm not sure to get the example, in the sense that it does not make sense to me... do you have a concrete example where you want to set two time limits? 
In your example, it feels that your new problem (50 steps limit) is treated as an infinite horizon problem when you bootstrap (because the agent never sees the real time limit). As in the infinite-horizon, the true horizon is dependent on the discount factor.

The only case where time-aware and infinite horizon is not equivalent for me is when you want your agent to do something special depending on the time, for instance a jump in the last steps.

> The agent is not aware of the timelimit - here the distinction doesn't matter, since the agent is unaware of either timelimit, and we can always safely bootstrap.

In that case, you are breaking Markov assumption...


> For the TD update,

My point was more to give a high-level reason of why we bootstrap or not, and dive into the details later.

Reply:
> Terminated could be true (for instance in case of a failure)

Ohh oops yes, I meant terminated cannot be True due to a time limit here (given this PR's meaning and usage of terminated) 

> I'm not sure to get the example, in the sense that it does not make sense to me... do you have a concrete example where you want to set two time limits?

I was imagining a scenario where the agent had an age that was very long (say 10k steps) which it's aware of, but for the purpose of diversifying we wish to restart the episode from when the agent is at t>>0, and let it learn for 50 steps at at time. But it occurs to me that the age would no longer be considered as a timelimit (since it's no longer 10k steps from the start of the episode), rather a regular termination condition. So, my point is invalid  :thinking: I think I also need to think more about this and what 'timelimit' means. I'll get back on the issue thread. 

> In that case, you are breaking Markov assumption...

Right, so if the agent has to optimize for a time-limit, it *has* to be time-aware to satisfy the markov assumption 

> My point was more to give a high-level reason of why we bootstrap or not, and dive into the details later.

Oh okay, that makes sense